File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/viewmodels/AccountSecurityViewModel.kt
// v1.2 Phase 4 — Settings → Multi-device + 2FA. Kotlin mirror of the
// iOS AccountSecurityViewModel. Drives:
//
//   1. Load the account-type badge (single vs multi-device).
//   2. Four-step enrollment (explainer → QR → sample code → recovery
//      codes display).
//   3. Disable an existing enrollment (Phase 3 endpoint; refused by
//      the Worker when other paired sessions still exist).
//
// IRK signatures for enroll-begin / enroll-confirm / disable are
// produced from the current Keystore IRK seed — same primitive
// ReplaceDeviceViewModel uses.

package com.flagshipserver.app.viewmodels

import androidx.lifecycle.ViewModel
import com.flagshipserver.app.api.FlagshipServerClient
import com.flagshipserver.app.api.TotpDisableRequest
import com.flagshipserver.app.api.TotpEnrollBeginRequest
import com.flagshipserver.app.api.TotpEnrollConfirmRequest
import com.flagshipserver.app.core.HexUtil
import com.flagshipserver.app.keystore.Keystore
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

sealed interface AccountSecurityPhase {
    data object Idle : AccountSecurityPhase
    data object Beginning : AccountSecurityPhase
    data class Staged(
        val secret: String,
        val otpauthUrl: String,
        val qrPngBase64: String,
        val issuer: String,
    ) : AccountSecurityPhase
    data object Confirming : AccountSecurityPhase
    data class Confirmed(
        val totpEnrolledAt: Long,
        val recoveryCodes: List<String>,
    ) : AccountSecurityPhase
    data object Disabling : AccountSecurityPhase
    data object Disabled : AccountSecurityPhase
    data class Failed(val message: String) : AccountSecurityPhase
}

class AccountSecurityViewModel(
    private val server: FlagshipServerClient,
    private val username: () -> String?,
) : ViewModel() {

    private val _phase = MutableStateFlow<AccountSecurityPhase>(AccountSecurityPhase.Idle)
    val phase: StateFlow<AccountSecurityPhase> = _phase.asStateFlow()

    /** Current account-type badge ("single" / "multi"). Null while
     *  loading or after a failure. */
    private val _accountType = MutableStateFlow<String?>(null)
    val accountType: StateFlow<String?> = _accountType.asStateFlow()

    private val _totpEnrolledAt = MutableStateFlow<Long?>(null)
    val totpEnrolledAt: StateFlow<Long?> = _totpEnrolledAt.asStateFlow()

    /** Convenience for the badge label/copy. */
    val isMultiDevice: Boolean get() = _accountType.value == "multi"

    suspend fun load() {
        val user = username()
        if (user.isNullOrEmpty()) {
            _accountType.value = null
            _totpEnrolledAt.value = null
            return
        }
        try {
            val rec = server.getUsernameRecord(user)
            _accountType.value = rec.accountType
            _totpEnrolledAt.value = rec.totpEnrolledAt
        } catch (_: Throwable) {
            _accountType.value = null
            _totpEnrolledAt.value = null
        }
    }

    suspend fun beginEnrollment() {
        val user = username()
        if (user.isNullOrEmpty()) {
            _phase.value = AccountSecurityPhase.Failed("No active account on this device.")
            return
        }
        _phase.value = AccountSecurityPhase.Beginning

        val issuedAt = System.currentTimeMillis()
        val signature = try {
            val signer = Keystore.deriveIRK("Enable multi-device + 2FA", Keystore.currentIrkVersion())
            signer.sign(canonicalEnrollBegin(user, issuedAt))
        } catch (e: Throwable) {
            _phase.value = AccountSecurityPhase.Failed("Couldn't sign enroll-begin: ${e.message}")
            return
        }

        try {
            val resp = server.totpEnrollBegin(
                username = user,
                body = TotpEnrollBeginRequest(
                    request = TotpEnrollBeginRequest.Inner(username = user, issuedAt = issuedAt),
                    signature = HexUtil.encode(signature),
                ),
            )
            _phase.value = AccountSecurityPhase.Staged(
                secret = resp.secret,
                otpauthUrl = resp.otpauthUrl,
                qrPngBase64 = resp.qrPngBase64,
                issuer = resp.issuer,
            )
        } catch (e: Throwable) {
            val msg = e.message.orEmpty()
            val friendly = if (msg.contains("503")) {
                "Multi-device + 2FA isn't configured on this server yet. Try again later."
            } else {
                "Couldn't start enrollment: $msg"
            }
            _phase.value = AccountSecurityPhase.Failed(friendly)
        }
    }

    suspend fun confirmEnrollment(sampleCode: String) {
        val trimmed = sampleCode.trim()
        if (trimmed.isEmpty()) {
            _phase.value = AccountSecurityPhase.Failed("Enter the 6-digit code from your authenticator app.")
            return
        }
        val user = username()
        if (user.isNullOrEmpty()) {
            _phase.value = AccountSecurityPhase.Failed("No active account on this device.")
            return
        }
        _phase.value = AccountSecurityPhase.Confirming

        val issuedAt = System.currentTimeMillis()
        val signature = try {
            val signer = Keystore.deriveIRK("Confirm 2FA code", Keystore.currentIrkVersion())
            signer.sign(canonicalEnrollConfirm(user, issuedAt))
        } catch (e: Throwable) {
            _phase.value = AccountSecurityPhase.Failed("Couldn't sign enroll-confirm: ${e.message}")
            return
        }

        try {
            val resp = server.totpEnrollConfirm(
                username = user,
                body = TotpEnrollConfirmRequest(
                    request = TotpEnrollConfirmRequest.Inner(username = user, issuedAt = issuedAt),
                    signature = HexUtil.encode(signature),
                    code = trimmed,
                ),
            )
            _accountType.value = resp.accountType
            _totpEnrolledAt.value = resp.totpEnrolledAt
            _phase.value = AccountSecurityPhase.Confirmed(
                totpEnrolledAt = resp.totpEnrolledAt,
                recoveryCodes = resp.recoveryCodes,
            )
        } catch (e: Throwable) {
            val msg = e.message.orEmpty()
            val friendly = if (msg.contains("401")) {
                "That code didn't match. Try again with a fresh code from your authenticator."
            } else {
                "Couldn't confirm: $msg"
            }
            _phase.value = AccountSecurityPhase.Failed(friendly)
        }
    }

    /** User has saved the recovery codes; dismiss the sheet and scrub
     *  the plaintexts from memory. The account-type / enrolled-at
     *  stays set since the Worker enrollment is committed. */
    fun dismissEnrollment() {
        _phase.value = AccountSecurityPhase.Idle
    }

    suspend fun disableEnrollment(code: String) {
        val trimmed = code.trim()
        if (trimmed.isEmpty()) {
            _phase.value = AccountSecurityPhase.Failed("Enter your 6-digit code or a recovery code to confirm.")
            return
        }
        val user = username()
        if (user.isNullOrEmpty()) {
            _phase.value = AccountSecurityPhase.Failed("No active account on this device.")
            return
        }
        _phase.value = AccountSecurityPhase.Disabling
        val issuedAt = System.currentTimeMillis()
        val signature = try {
            val signer = Keystore.deriveIRK("Disable multi-device + 2FA", Keystore.currentIrkVersion())
            signer.sign(canonicalDisable(user, issuedAt))
        } catch (e: Throwable) {
            _phase.value = AccountSecurityPhase.Failed("Couldn't sign disable: ${e.message}")
            return
        }
        try {
            val resp = server.totpDisable(
                username = user,
                body = TotpDisableRequest(
                    request = TotpDisableRequest.Inner(username = user, issuedAt = issuedAt),
                    signature = HexUtil.encode(signature),
                    code = trimmed,
                ),
            )
            _accountType.value = resp.accountType
            _totpEnrolledAt.value = null
            _phase.value = AccountSecurityPhase.Disabled
        } catch (e: Throwable) {
            val msg = e.message.orEmpty()
            val friendly = when {
                msg.contains("401") -> "That code didn't match. Try a fresh code from your authenticator."
                msg.contains("409") -> "Disable refused — other devices are still trusted on this account. Disconnect them first."
                else -> "Couldn't disable: $msg"
            }
            _phase.value = AccountSecurityPhase.Failed(friendly)
        }
    }

    companion object {
        /** Mirrors packages/protocol/src/auth.ts canonicalTotpEnrollBegin. */
        fun canonicalEnrollBegin(username: String, issuedAt: Long): ByteArray =
            "flagship/totp-enroll-begin/v1|$username|$issuedAt".toByteArray(Charsets.UTF_8)

        fun canonicalEnrollConfirm(username: String, issuedAt: Long): ByteArray =
            "flagship/totp-enroll-confirm/v1|$username|$issuedAt".toByteArray(Charsets.UTF_8)

        fun canonicalDisable(username: String, issuedAt: Long): ByteArray =
            "flagship/totp-disable/v1|$username|$issuedAt".toByteArray(Charsets.UTF_8)
    }
}

File: apps/mobile/ios/Sources/FlagshipUI/ViewModels/AccountSecurityViewModel.swift
import Foundation
import Observation
import Flagship
import FlagshipAPI
import FlagshipCore

/// v1.2 Phase 4 — drives the Settings → Multi-device + 2FA flow:
///
///   1. Load the account-type badge state from
///      `GET /api/users/:u` (single vs multi-device).
///   2. Step the user through the four-step enroll sheet (explainer
///      → QR + secret → sample code → recovery codes).
///   3. Optionally disable an existing enrollment (Phase 3 endpoint;
///      gated by the Worker on single-device-count).
///
/// The IRK signature for enroll-begin / enroll-confirm / disable is
/// produced by Keystore.deriveIRK at the current IRK version — same
/// shape as ReplaceDeviceViewModel.
///
/// Recovery codes (the ten plaintext strings returned ONCE on
/// enroll-confirm) are kept only in this view-model's `.confirmed`
/// payload — they leave the device the moment the user dismisses
/// the codes screen, which is gated behind an explicit
/// "I've saved these" tap.
@Observable
@MainActor
public final class AccountSecurityViewModel {

    /// Discriminated state machine for the enrollment sheet.
    public enum Phase: Equatable, Sendable {
        case idle
        case beginning           // signing + POST /enroll-begin in flight
        case staged(StagedSecret) // got QR + base32 secret; awaiting the sample code
        case confirming          // POST /enroll-confirm in flight
        case confirmed(EnrollmentResult)
        case disabling
        case disabled
        case failed(String)
    }

    public struct StagedSecret: Equatable, Sendable {
        public let secret: String
        public let otpauthUrl: String
        public let qrPngBase64: String
        public let issuer: String
        public init(secret: String, otpauthUrl: String, qrPngBase64: String, issuer: String) {
            self.secret = secret
            self.otpauthUrl = otpauthUrl
            self.qrPngBase64 = qrPngBase64
            self.issuer = issuer
        }
    }

    public struct EnrollmentResult: Equatable, Sendable {
        public let totpEnrolledAt: Int64
        public let recoveryCodes: [String]
        public init(totpEnrolledAt: Int64, recoveryCodes: [String]) {
            self.totpEnrolledAt = totpEnrolledAt
            self.recoveryCodes = recoveryCodes
        }
    }

    public private(set) var phase: Phase = .idle

    /// Current account-type badge state. Refreshed by `load()`. Pre-
    /// migration accounts (no row in /users/:u) surface as nil so the
    /// Settings screen can render a graceful placeholder.
    public private(set) var accountType: String?
    public private(set) var totpEnrolledAt: Int64?

    private let server: FlagshipServerClient
    private let username: @MainActor () -> String?

    public init(server: FlagshipServerClient, username: @MainActor @escaping () -> String?) {
        self.server = server
        self.username = username
    }

    /// Build the watch-delegate ("Quick approve from Apple Watch") toggle VM
    /// from the same server + username injection, so the Account-security
    /// screen can host the toggle without the app layer wiring a second
    /// dependency graph. Keystore-backed crypto/local-store use their
    /// defaults (this runs on a real device / the sim Keychain fallback).
    public func makeWatchDelegateViewModel() -> WatchDelegateViewModel {
        WatchDelegateViewModel(server: server, username: username)
    }

    /// Read the current account-type from the Worker. Idempotent;
    /// safe to call on every Settings screen open.
    public func load() async {
        guard let user = username(), !user.isEmpty else {
            accountType = nil
            totpEnrolledAt = nil
            return
        }
        do {
            let resp = try await server.getUsernameRecord(username: user)
            accountType = resp.accountType
            totpEnrolledAt = resp.totpEnrolledAt
        } catch {
            // Non-fatal — Settings still renders, just without the
            // badge until the next refresh.
            accountType = nil
            totpEnrolledAt = nil
        }
    }

    /// Convenience for the badge label/copy. Read-only; doesn't
    /// branch the UI past the displayed string.
    public var isMultiDevice: Bool { accountType == "multi" }

    /// Step 1→2: sign enroll-begin and ask the Worker to stage a
    /// secret. On success drops into `.staged`, which the sheet
    /// renders as the QR + manual-key + "I've added it" step.
    public func beginEnrollment() async {
        guard let user = username(), !user.isEmpty else {
            phase = .failed("No active account on this device.")
            return
        }
        phase = .beginning
        let issuedAt = Int64(Date().timeIntervalSince1970 * 1000)
        let irk: Curve25519.Signing.PrivateKey
        do {
            irk = try await Keystore.deriveIRK(
                reason: "Enable multi-device + 2FA",
                version: Keystore.currentIrkVersion()
            )
        } catch {
            phase = .failed("Couldn't access your account keys: \(error.localizedDescription)")
            return
        }
        let canonical = TotpEnrollBeginCanonical.bytes(username: user, issuedAt: issuedAt)
        let signature: Data
        do {
            signature = try irk.signature(for: canonical)
        } catch {
            phase = .failed("Couldn't sign enroll-begin: \(error.localizedDescription)")
            return
        }
        do {
            let resp = try await server.totpEnrollBegin(
                username: user,
                body: TotpEnrollBeginRequest(
                    request: .init(username: user, issuedAt: issuedAt),
                    signature: HexUtil.encode(signature)
                )
            )
            phase = .staged(StagedSecret(
                secret: resp.secret,
                otpauthUrl: resp.otpauthUrl,
                qrPngBase64: resp.qrPngBase64,
                issuer: resp.issuer
            ))
        } catch ScreensClientError.http(let status, _) where status == 503 {
            phase = .failed("Multi-device + 2FA isn't configured on this server yet. Try again later.")
        } catch {
            phase = .failed("Couldn't start enrollment: \(error.localizedDescription)")
        }
    }

    /// Step 3→4: submit the user-entered sample code; on success the
    /// account flips to `'multi'` and 10 recovery codes are returned
    /// once.
    public func confirmEnrollment(sampleCode: String) async {
        let trimmed = sampleCode.trimmingCharacters(in: .whitespaces)
        guard !trimmed.isEmpty else {
            phase = .failed("Enter the 6-digit code from your authenticator app.")
            return
        }
        guard let user = username(), !user.isEmpty else {
            phase = .failed("No active account on this device.")
            return
        }
        // Keep the staged secret around so a sad-path retry can hop
        // back into .staged without re-deriving.
        let priorStaged: StagedSecret? = {
            if case .staged(let s) = phase { return s }
            return nil
        }()
        phase = .confirming

        let issuedAt = Int64(Date().timeIntervalSince1970 * 1000)
        let irk: Curve25519.Signing.PrivateKey
        do {
            irk = try await Keystore.deriveIRK(
                reason: "Confirm 2FA code",
                version: Keystore.currentIrkVersion()
            )
        } catch {
            phase = .failed("Couldn't access your account keys: \(error.localizedDescription)")
            return
        }
        let canonical = TotpEnrollConfirmCanonical.bytes(username: user, issuedAt: issuedAt)
        let signature: Data
        do {
            signature = try irk.signature(for: canonical)
        } catch {
            phase = .failed("Couldn't sign enroll-confirm: \(error.localizedDescription)")
            return
        }
        do {
            let resp = try await server.totpEnrollConfirm(
                username: user,
                body: TotpEnrollConfirmRequest(
                    request: .init(username: user, issuedAt: issuedAt),
                    signature: HexUtil.encode(signature),
                    code: trimmed
                )
            )
            accountType = resp.accountType
            totpEnrolledAt = resp.totpEnrolledAt
            phase = .confirmed(EnrollmentResult(
                totpEnrolledAt: resp.totpEnrolledAt,
                recoveryCodes: resp.recoveryCodes
            ))
        } catch ScreensClientError.http(let status, _) where status == 401 {
            // Bounce back to .staged so the user can retry without
            // restarting the whole flow.
            if let staged = priorStaged {
                phase = .staged(staged)
            }
            // Surface the mismatch as a secondary failed state — the
            // UI re-reads `phase` and shows an inline error beside
            // the input.
            phase = .failed("That code didn't match. Try again with a fresh code from your authenticator.")
        } catch {
            phase = .failed("Couldn't confirm: \(error.localizedDescription)")
        }
    }

    /// User has saved the recovery codes; dismiss the sheet and
    /// scrub them from memory.
    public func dismissEnrollment() {
        phase = .idle
    }

    /// Disable a previously-enrolled 2FA. The user provides a live
    /// 6-digit code (or recovery code). Refused if other paired
    /// sessions exist (Worker enforces).
    public func disableEnrollment(code: String) async {
        let trimmed = code.trimmingCharacters(in: .whitespaces)
        guard !trimmed.isEmpty else {
            phase = .failed("Enter your 6-digit code or a recovery code to confirm.")
            return
        }
        guard let user = username(), !user.isEmpty else {
            phase = .failed("No active account on this device.")
            return
        }
        phase = .disabling
        let issuedAt = Int64(Date().timeIntervalSince1970 * 1000)
        let irk: Curve25519.Signing.PrivateKey
        do {
            irk = try await Keystore.deriveIRK(
                reason: "Disable multi-device + 2FA",
                version: Keystore.currentIrkVersion()
            )
        } catch {
            phase = .failed("Couldn't access your account keys: \(error.localizedDescription)")
            return
        }
        let canonical = TotpDisableCanonical.bytes(username: user, issuedAt: issuedAt)
        let signature: Data
        do {
            signature = try irk.signature(for: canonical)
        } catch {
            phase = .failed("Couldn't sign disable: \(error.localizedDescription)")
            return
        }
        do {
            let resp = try await server.totpDisable(
                username: user,
                body: TotpDisableRequest(
                    request: .init(username: user, issuedAt: issuedAt),
                    signature: HexUtil.encode(signature),
                    code: trimmed
                )
            )
            accountType = resp.accountType
            totpEnrolledAt = nil
            phase = .disabled
        } catch ScreensClientError.http(let status, _) where status == 401 {
            phase = .failed("That code didn't match. Try a fresh code from your authenticator.")
        } catch ScreensClientError.http(let status, _) where status == 409 {
            phase = .failed("Disable refused — other devices are still trusted on this account. Disconnect them first.")
        } catch {
            phase = .failed("Couldn't disable: \(error.localizedDescription)")
        }
    }
}

#if canImport(CryptoKit)
import CryptoKit
#endif

/// Canonical-bytes helpers — mirror @flagship/protocol's
/// `canonicalTotpEnrollBegin` / `canonicalTotpEnrollConfirm` /
/// `canonicalTotpDisable`. Defined locally so the VM doesn't depend
/// on the JS-side protocol package (the iOS app already mirrors the
/// canonical-bytes everywhere else through these tiny helpers).
enum TotpEnrollBeginCanonical {
    static func bytes(username: String, issuedAt: Int64) -> Data {
        Data("flagship/totp-enroll-begin/v1|\(username)|\(issuedAt)".utf8)
    }
}
enum TotpEnrollConfirmCanonical {
    static func bytes(username: String, issuedAt: Int64) -> Data {
        Data("flagship/totp-enroll-confirm/v1|\(username)|\(issuedAt)".utf8)
    }
}
enum TotpDisableCanonical {
    static func bytes(username: String, issuedAt: Int64) -> Data {
        Data("flagship/totp-disable/v1|\(username)|\(issuedAt)".utf8)
    }
}
